### PR TITLE
refactor: clean up retrieval from data transfer item

### PR DESCRIPTION
**Pull request by jonkoops**
**What kind of change does this PR introduce?**
- [ ] Bug Fix
- [ ] Feature
- [x] Refactoring
- [ ] Style
- [ ] Build
- [ ] Chore
- [ ] Documentation
- [ ] CI

**Did you add tests for your changes?**
- [x] Yes, my code is well tested
- [ ] Not relevant

**If relevant, did you update the documentation?**
- [ ] Yes, I've updated the documentation
- [x] Not relevant

**Summary**
Cleans up the code in the `fromDataTransferItem()` function so that it is fully type-safe and uses `await` everywhere for Promises.

**Does this PR introduce a breaking change?**
No

**Other information**
Not relevant

### CODE REVIEW

**Comment by jonkoops** [comment]
This PR builds off the work from #112, so that should be merged first so this can be rebased on top.

**Comment by jonkoops** [comment]
Re-based the PR and cleaned it up a little, I think we're good to go like this. @rolandjitsu can I ask you for a review?

File: src/file-selector.ts
Comment (rolandjitsu) [inline src/file-selector.ts:127]: Make a func here to return this err (same err in 2 places)? Or make a custom error object with the same msg?

File: src/file-selector.ts
Comment (rolandjitsu) [reply src/file-selector.ts:121]: The intent was to see what exactly is not a file ... so it might make sense to make a custom error object that can embed the `item` which could later be inspected if they look at the error?

Comment (jonkoops) [reply src/file-selector.ts:121]: We could do that yeah, proper custom error types [are a bit involved](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error#custom_error_types) though, so I'd rather avoid it. I also don't know how much value there would be in having the `item` available here.

File: src/file-selector.ts
Comment (jonkoops) [reply src/file-selector.ts:127]: Could do that yeah, but seems a bit overkill. I generally use the [Rule of Three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) for this kind of stuff.

File: src/file-selector.ts
Comment (MathiasWP) [inline src/file-selector.ts:116]: Should an error be thrown here? Throwing this error would still cause https://github.com/react-dropzone/file-selector/issues/120 to be reported, which is not desired.

Comment (jonkoops) [reply src/file-selector.ts:116]: Not entirely sure, but we should already be [filtering out](https://github.com/react-dropzone/file-selector/blob/1a997bd800fcfa60d8523a69f4b361fee99965b9/src/file-selector.ts#L63) anything that is not a file. I am not really sure how we could get a handle back here, and it not be a file, but this is handling the edge cases we never expect to encounter.

Comment (jonkoops) [reply src/file-selector.ts:116]: Gonna resolve this discussion, let's use your issue instead.

**Comment by jonkoops** [comment]
I'll put this PR on hold until we merged the backport under #122, as this is likely to create conflicts.

**Comment by jonkoops** [comment]
@rolandjitsu I've rebased this PR on top of the fixes we backported from `master`, could I ask you for another pass?

Since the backports included converting the method to use `await` that enhancement no longer applies here, but the changes still include some additional type-safety fixes and improves the error messages, along with splitting up some of the code a bit more, so I think it is still relevant to land this.

**Comment by jonkoops** [comment]
As for the advanced errors with details about the handle, I think we could handle (no pun intended) that as a separate feature. Since the original error messages are kinda broken, and I have not seen anyone complain about it I think we can safely assume it is not that important to users, or they just don't get these messages often (or at all).

File: src/file-selector.ts
Comment (rolandjitsu) [inline src/file-selector.ts:131]: I mean, it has no functional value, but the original message change wasn't necessary 😉

File: src/file-selector.ts
Comment (rolandjitsu) [inline src/file-selector.ts:113]: I'm really not a fan of these empty lines. It doesn't really improve readability, it just makes adds more eye movement 😆 Please remove them.

File: src/file-selector.ts
Comment (rolandjitsu) [inline src/file-selector.ts:112]: Coming from Go, these explicit variable names feel so long 😆 When a simple `f` could do.

File: src/file-selector.ts
Comment (rolandjitsu) [inline src/file-selector.ts:140]: Since where this function is used, `null` is not explicitly checked, a simple:
```suggestion
    return;
```

would also work since `if (undefined)` is the same as `if (null)`. 

You can probably tell that I'm not a fan of adding unnecessary code when functionally a shorter version does the same (and there's really no gains in improving readability either).

File: src/file-selector.ts
Comment (rolandjitsu) [inline src/file-selector.ts:151]: `isFileHandle` can and should just handle the `null` case.

File: src/file-selector.ts
Comment (rolandjitsu) [inline src/file-selector.ts:218]: Would this cause issues in downstream packages if they have different declarations? I'm not up to date with how the `declare global` works and if it crosses the boundary b/t packages (if 2 packages declare the same in the global namespace).

File: src/file-selector.ts
Comment (jonkoops) [reply src/file-selector.ts:131]: Fair, I will restore the original. I think I changed it because it did not make sense in some initial refactoring I did, but it's pretty much the same after re-basing.

File: src/file-selector.ts
Comment (jonkoops) [reply src/file-selector.ts:112]: I prefer it explicit to be honest, more often than not when a block gets more complicated this starts feeling like you don't know what you are operating on any longer. E.g. below we also have a variable `file`, so `f` gets a bit muddled. That said, I could solve this much nicer if we had basic pattern matching in JS.

File: src/file-selector.ts
Comment (jonkoops) [reply src/file-selector.ts:140]: In this case I just follow what the API returns here as well, I don't like adding polymorphism where it is not needed. In most cases `null` is meant as a 'this operation yields an empty result' and `undefined` means 'this is void' (no value is ever returned). You'll see this convention in most APIs that exist in the standard API of JavaScript runtimes.

File: src/file-selector.ts
Comment (jonkoops) [reply src/file-selector.ts:151]: I don't think it belongs there, `isFileHandle()` is meant as a simple utility to check if a `FileSystemHandle` is a `FileSystemFileHandle` or not. Whether the variable is `null` is an entirely different check, e.g. none of the other functions check for nullish values either (e.g. `isDirectoryEntry()`).

**Comment by jonkoops** [comment]
@rolandjitsu did another pass and processed your comments, could I ask you for another review?

**Review by rolandjitsu — APPROVED** [review]

**Comment by rolandjitsu** [comment]
> @rolandjitsu did another pass and processed your comments, could I ask you for another review?

Sorry for the long wait. Got a bit busy at work.

**Comment by jonkoops** [comment]
No worries, I know how it is :)

**Pull request merged**
